pkg/operator/controller/configurable-route/controller_test.go

```diff
@@ -0,0 +1,899 @@
+				ObjectMeta: v1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "openshift-ingress-operator",
+				},
+				Spec: configv1.IngressSpec{
+					ComponentRoutes: []configv1.ComponentRouteSpec{
+						{
+							Name:      "foo",
+							Namespace: "openshift-config",
+							ServingCertKeyPairSecret: configv1.SecretNameReference{
+								Name: "foo-cert",
+							},
+						},
+					},
+				},
+				Status: configv1.IngressStatus{
+					ComponentRoutes: []configv1.ComponentRouteStatus{
+						{
+							Name:      "foo",
+							Namespace: "openshift-config",
+							ConsumingUsers: []configv1.ConsumingUser{
+								"system:serviceaccount:openshift-config:foo-cert",
+							},
+						},
+					},
+				},
+			},
+			roles: rbacv1.RoleList{
+				Items: []rbacv1.Role{
+					*newRoleForServiceCert("bar", "openshift-config", "bar-hash", "foo-cert"),
+				},
+			},
+			expectedRoles: rbacv1.RoleList{
+				Items: []rbacv1.Role{
+					*newRoleForServiceCertWithGeneratedHash("foo", "openshift-config", "foo-cert"),
+				},
+			},
+			expectedRoleBindings: rbacv1.RoleBindingList{
+				Items: []rbacv1.RoleBinding{
+					*newRoleBindingForServiceCertWithGeneratedHash("foo", "openshift-config", "foo", "foo-cert", "openshift-config"),
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			r := newFakeReconciler()
+			if tc.ingressConfig != nil {
+				if err := r.client.Create(context.Background(), tc.ingressConfig); err != nil {
+					t.Errorf("error creating ingress config: %v", err)
+				}
+			}
+
+			if _, err := r.Reconcile(context.Background(), tc.request); err == nil && tc.expectError {
+				t.Errorf("expected error, got no error")
+			} else if err != nil && !tc.expectError {
+				t.Errorf("did not expected error: %v", err)
+			} else {
+				actualRoles := &rbacv1.RoleList{}
+				if err := r.client.List(context.Background(), actualRoles); err == nil {
+					for _, expectedRole := range tc.expectedRoles.Items {
+						// Find the actual role that we should compare status with this expected role
+						var actualRole rbacv1.Role
+						for _, role := range actualRoles.Items {
+							// TRICKY: Because ensureServiceCertKeyPairSecretRole *sometimes* generates a name for the
+							//         role, we need to compare both name and generated name.
+							if role.GenerateName == expectedRole.GenerateName || role.Name == expectedRole.Name {
+								actualRole = role
+								expectedRole.Name = role.Name
+							}
+						}
+						if assert.NotEmptyf(t, actualRole, fmt.Sprintf("Could not find expected role: %s/%s", expectedRole.Namespace, expectedRole.Name)) {
+							// Value that we want to ignore in the comparison.
+							actualRole.ResourceVersion = ""
+							actualRole.GenerateName = ""
+							expectedRole.GenerateName = ""
+							assert.Equal(t, expectedRole, actualRole)
+						}
+					}
+				} else {
+					t.Errorf("error retrieving roles from client: %v", err)
+				}
```

Comment:
Can you encapsulate these sections around the `List` calls to the api and make them helper functions?

Reply:
Good idea. Done.